contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,565 @@ const (
+For this use case, the user's specs are unchanged. Only one change
+to the StorageClass is needed to enable delayed volume binding.
+
+1. Admin sets up StorageClass, setting up delayed volume binding.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard
+provisioner: kubernetes.io/gce-pd
+volumeBindingMode: WaitForFirstConsumer
+parameters:
+  type: pd-standard
+```
+2. Admin launches provisioner.  For in-tree plugins, nothing needs to be done.
+3. User creates PVC. Nothing changes in the spec, although now the PVC won't be
+   immediately bound.
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: my-pvc
+spec:
+  storageClassName: standard
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 100Gi
+```
+4. User creates Pod. Nothing changes in the spec.
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-pod
+spec:
+  containers:
+  ...
+  volumes:
+  - name: my-vol
+    persistentVolumeClaim:
+      claimName: my-pvc
+```
+5. Scheduler picks a node that can satisfy the Pod and passes it to the
+   provisioner.
+6. Provisioner dynamically provisions a PV that can be accessed from
+   that node.
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: standard
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+```
+7. Pod gets scheduled to the node.
+
+
+### Restricting Topology
+For the common use case, volumes will be provisioned in whatever topology domain
+the scheduler has decided is best to run the workload. Users may impose further
+restrictions by setting label/node selectors, and pod affinity/anti-affinity
+policies on their Pods. All those policies will be taken into account when
+dynamically provisioning a volume.
+
+While less common, administrators may want to further restrict what topology
+domains are available to a StorageClass. To support these administrator
+policies, an AllowedTopology field can also be specified in the
+StorageClass to restrict the topology domains for dynamic provisioning.
+This is not expected to be a common use case, and there are some caveats,
+described below.
+
+```
+type StorageClass struct {
+    ...
+
+    AllowedTopologies *VolumeProvisioningTopologies
+}
+
+type VolumeProvisioningTopologies struct {
+    // Volume must be provisioned from one of the required topologies.
+    // These VolumeTopology requirements are ORed.
+    RequiredTopologies []VolumeTopology
+}
+
+type VolumeTopology struct{
+    // The provisioned volume must support all the specified node labels.
+    // Each entry in the map is ANDed.
+    MatchNodeLabels map[string]string
+}
+```
```

Comment:
> ```
> StorageClass:
>  topologyConstraints:
>    - zone in [a, b]
>      rack in ["1", "2"]
> ```
Wouldn't this allow "zone=b,rack=1"? To entirely capture the original topology, you'd need something like:
```
StorageClass:
  topologyConstraints:
    - zone in [a]
      rack in ["1"]
    - zone in [b]
      rack in ["2"]
```

One inconvenience with the selector approach is there are multiple ways to specify certain kinds of topology. For example, the following 3 are equivalent:
```
StorageClass:
  topologyConstraints:
    - zone in [a]
      rack in ["1"]
    - zone in [b]
      rack in ["1"]
    - zone in [b]
      rack in ["2"]
```
```
StorageClass:
  topologyConstraints:
    - zone in [a,b]
      rack in ["1"]
    - zone in [b]
      rack in ["2"]
```
```
StorageClass:
  topologyConstraints:
    - zone in [a]
      rack in ["1"]
    - zone in [b]
      rack in ["1", "2"]
```

I could potentially see users being confused which form Kubernetes expects, when in reality either of them work. To me it seems better to have one layer of ORs instead of two.

Reply:
My updated example [here](https://github.com/kubernetes/community/pull/1857/files#diff-9d29da618fae1244e58d194d3d70e2adR554) correctly represents (one way) the example topology.  I think having multiple representations is fine as long as they evaluate the same, and it is consistent with how selectors are specified for other Kubernetes objects.